### PR TITLE
Add EffOpenBook

### PR DESCRIPTION
**Pull request by OfficialDonut**
### Description
Added open book effect to open written books to players.

---
**Target Minecraft Versions:** any
**Requirements:** no
**Related Issues:** no

### CODE REVIEW

File: src/main/java/ch/njol/skript/effects/EffOpenBook.java [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+	}
+	
+	@SuppressWarnings("null")
+	private Expression<ItemType> book;
+	@SuppressWarnings("null")
+	private Expression<Player> players;
+	
+	@SuppressWarnings({"unchecked", "null"})
+	@Override
+	public boolean init(final Expression<?>[] exprs, final int matchedPattern, final Kleenean isDelayed, final SkriptParser.ParseResult parseResult) {
```
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/effects/EffOpenBook.java:55]: You should import ParseResult directly

File: src/main/java/ch/njol/skript/effects/EffOpenBook.java [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+				}
+			}
+		}
+	}
+	
+	@SuppressWarnings("null")
+	@Override
+	public String toString(@Nullable Event e, boolean debug) {
+		return "open book " + book.toString(e, debug) + " to " + players.toString(e, debug);
+	}
```
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/effects/EffOpenBook.java:78]: Nitpicky but could you add a space after here
```suggestion
	}

```

File: src/main/java/ch/njol/skript/effects/EffOpenBook.java [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+import ch.njol.util.Kleenean;
+
+@Name("Open Book")
+@Description("Opens a written book to a player.")
+@Examples("open book player's tool to player")
+@Since("INSERT VERSION")
+public class EffOpenBook extends Effect {
+	
+	static {
+		Skript.registerEffect(EffOpenBook.class, "(open|show) book %itemtype% to %players%");
```
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/effects/EffOpenBook.java:45]:
```suggestion
		Skript.registerEffect(EffOpenBook.class, "(open|show) book %itemtype% (to|for) %players%");
```

**Review by APickledWalrus — CHANGES_REQUESTED** [review]

**Review by Pikachu920 — CHANGES_REQUESTED** [review]

File: .gitattributes [hunk 4]
```diff
@@ -2,7 +2,7 @@
         #
         # Handle line endings automatically for files detected as text
         # and leave all files detected as binary untouched.
-        * text=auto
```
[the remark this answers is not in the document]
Comment (ShaneBeee) [reply .gitattributes]: that could be due to it not being on the 2.5 branch, just updated it, should go away

File: src/main/java/ch/njol/skript/effects/EffOpenBook.java [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+		book = (Expression<ItemType>) exprs[0];
+		players = (Expression<Player>) exprs[1];
+		return true;
+	}
+	
+	@Override
+	protected void execute(final Event e) {
+		@Nullable ItemType itemType = book.getSingle(e);
+		if (itemType != null && itemType.getMaterial() == Material.WRITTEN_BOOK) {
+			@Nullable ItemStack itemStack = itemType.getRandom();
```
[the remark this answers is not in the document]
Comment (OfficialDonut) [reply src/main/java/ch/njol/skript/effects/EffOpenBook.java:65]: It had a warning
```
Expression 'book.getSingle(e)' might evaluate to null but is assigned to a variable that is annotated with @NotNull
```

**Review by APickledWalrus — DISMISSED** [review]
LGTM

**Review by ShaneBeee — APPROVED** [review]

File: src/main/java/ch/njol/skript/effects/EffOpenBook.java [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+	public boolean init(final Expression<?>[] exprs, final int matchedPattern, final Kleenean isDelayed, final ParseResult parseResult) {
+		book = (Expression<ItemType>) exprs[0];
+		players = (Expression<Player>) exprs[1];
+		return true;
+	}
+	
+	@Override
+	protected void execute(final Event e) {
+		ItemType itemType = book.getSingle(e);
+		if (itemType != null && itemType.getMaterial() == Material.WRITTEN_BOOK) {
```
Comment (ShaneBeee) [inline src/main/java/ch/njol/skript/effects/EffOpenBook.java:64]: In the event a Bukkit material is changed or what not, an alias should be used here instead, here is an example of another class with books doing it:
https://github.com/SkriptLang/Skript/blob/99af0a13a7cefb0e6ed890e775093a51fc3802bf/src/main/java/ch/njol/skript/expressions/ExprBookTitle.java#L45
https://github.com/SkriptLang/Skript/blob/99af0a13a7cefb0e6ed890e775093a51fc3802bf/src/main/java/ch/njol/skript/expressions/ExprBookTitle.java#L54-L59

in this case you could use the ItemType#equals method instead before getting the item stack.

**Review by ShaneBeee — CHANGES_REQUESTED** [review]
small change.

**Review by ShaneBeee — APPROVED** [review]

**Review by APickledWalrus — APPROVED** [review]

**Pull request merged**